underfs/hdfs/src/main/java/alluxio/underfs/hdfs/HdfsPositionedUnderFileInputStream.java

```diff
@@ -61,22 +71,46 @@ public int read(byte[] b) throws IOException {
     return read(b, 0, b.length);
   }
 
+  /**
+   * If there has been a number of sequential reads in a row,
+   * we move to regular buffered reads.
+   */
   @Override
   public int read(byte[] buffer, int offset, int length) throws IOException {
-    int bytesRead = ((FSDataInputStream) in).read(mPos, buffer, offset, length);
+    int bytesRead;
+    if (mSequentialReadCount >= SEQUENTIAL_READ_LIMIT) {
+      ((FSDataInputStream) in).seek(mPos);
+      mSequentialReadCount = 0;
+    }
+    if (mPos == ((FSDataInputStream) in).getPos()) {
+      // same position, use buffered reads as default
+      bytesRead = in.read(buffer, offset, length);
+    } else {
+      bytesRead = ((FSDataInputStream) in).read(mPos, buffer, offset, length);
+    }
     if (bytesRead > 0) {
       mPos += bytesRead;
+      mSequentialReadCount++;
     }
     return bytesRead;
   }
 
   @Override
   public void seek(long position) throws IOException {
+    if (position < mPos || position - mPos > MOVEMENT_LIMIT) {
+      mSequentialReadCount = 0;
+    }
     mPos = position;
   }
 
   @Override
   public long skip(long n) throws IOException {
+    if (n <= 0) {
+      return 0;
+    }
+    if (n > MOVEMENT_LIMIT) {
+      mSequentialReadCount = 0;
+    }
```

Comment:
We still need to skip in the stream if we are in the sequential read mode and n is less than the movement limit?

Reply:
it is not necessary, because our mPos is the ground truth. As long as we maintain mPos, we can delay any seek or skip until when we actually read.

Reply:
In that case, a read pattern like `read, read, read, skip(1), read, skip(1), read` will cause us to lose the sequential read mode at the last read. Is this intended?

Reply:
yes, fixed. it became obvious once i started writing the test.